DigitalLearningSolutions.Web/ServiceFilter/ValidateAllowedDlsSubApplication.cs

```diff
@@ -33,42 +37,83 @@ public async void OnActionExecuting(ActionExecutingContext context)
             var user = context.HttpContext.User;
             if (!user.Identity.IsAuthenticated)
             {
-                RedirectToLogin(context);
+                SetLoginRedirectResult(context);
                 return;
             }
 
-            var dlsSubApplicationParameterName = context.ActionDescriptor.Parameters?
+            if (!TryGetDlsSubApplication(context, out var application, out var applicationParameterName))
+            {
+                return;
+            }
+
+            if (await ApplicationIsInaccessibleByPage(context, application))
+            {
+                return;
+            }
+
+            ValidateUserHasPermissionForApplication(context, user, application, applicationParameterName!);
+        }
+
+        private bool TryGetDlsSubApplication(
+            ActionExecutingContext context,
+            out DlsSubApplication? application,
+            out string? dlsSubApplicationParameterName
+        )
+        {
+            dlsSubApplicationParameterName = context.ActionDescriptor.Parameters?
                 .FirstOrDefault(x => x.ParameterType == typeof(DlsSubApplication))?.Name;
 
-            if (!string.IsNullOrEmpty(dlsSubApplicationParameterName) &&
-                HasModelBindingError(context, dlsSubApplicationParameterName!))
+            var actionDoesNotRequireApplication = dlsSubApplicationParameterName == null;
+            if (actionDoesNotRequireApplication)
+            {
+                application = null;
+                return false;
+            }
+
+            if (HasModelBindingError(context, dlsSubApplicationParameterName!))
             {
                 SetNotFoundResult(context);
-                return;
+                application = null;
+                return false;
             }
 
-            var application = dlsSubApplicationParameterName == null
-                ? null
-                : GetDlsSubApplicationFromContext(context, dlsSubApplicationParameterName);
+            application = GetDlsSubApplicationFromContext(context, dlsSubApplicationParameterName!);
+            return true;
+        }
 
+        private async Task<bool> ApplicationIsInaccessibleByPage(

```

Comment:
This name makes the method seem pure, but it actually sets the response.
Can we change the name to show that?